pkg/api/bzz.go

```diff
@@ -5,34 +5,234 @@
+				s.logger.Debugf("bzz upload file: increment tag: %v", err)
+				s.logger.Error("bzz upload file: increment tag")
+				jsonhttp.InternalServerError(w, "increment tag")
+				return
+			}
+		}
+	}
+
+	// Add the tag to the context
+	ctx := sctx.SetTag(r.Context(), tag)
+
+	fileName = r.URL.Query().Get("name")
+	contentLength = r.Header.Get("Content-Length")
+	reader = r.Body
+
+	if contentLength != "" {
+		fileSize, err = strconv.ParseUint(contentLength, 10, 64)
+		if err != nil {
+			logger.Debugf("bzz upload file: content length, file %q: %v", fileName, err)
+			logger.Errorf("bzz upload file: content length, file %q", fileName)
+			jsonhttp.BadRequest(w, "invalid content length header")
+			return
+		}
+	} else {
+		// copy the part to a tmp file to get its size
+		tmp, err := ioutil.TempFile("", "bee-multipart")
+		if err != nil {
+			logger.Debugf("bzz upload file: create temporary file: %v", err)
+			logger.Errorf("bzz upload file: create temporary file")
+			jsonhttp.InternalServerError(w, nil)
+			return
+		}
+		defer os.Remove(tmp.Name())
+		defer tmp.Close()
+		n, err := io.Copy(tmp, reader)
+		if err != nil {
+			logger.Debugf("bzz upload file: write temporary file: %v", err)
+			logger.Error("bzz upload file: write temporary file")
+			jsonhttp.InternalServerError(w, nil)
+			return
+		}
+		if _, err := tmp.Seek(0, io.SeekStart); err != nil {
+			logger.Debugf("bzz upload file: seek to beginning of temporary file: %v", err)
+			logger.Error("bzz upload file: seek to beginning of temporary file")
+			jsonhttp.InternalServerError(w, nil)
+			return
+		}
+		fileSize = uint64(n)
+		reader = tmp
+	}
+
+	p := requestPipelineFn(s.storer, r)
+
+	// first store the file and get its reference
+	fr, err := p(ctx, reader, int64(fileSize))
+	if err != nil {
+		logger.Debugf("bzz upload file: file store, file %q: %v", fileName, err)
+		logger.Errorf("bzz upload file: file store, file %q", fileName)
+		jsonhttp.InternalServerError(w, "could not store file data")
+		return
+	}
+
+	// If filename is still empty, use the file hash as the filename
+	if fileName == "" {
+		fileName = fr.String()
+	}
+
+	encrypt := requestEncrypt(r)
+	l := loadsave.New(s.storer, requestModePut(r), encrypt)
+
+	fileMtdt := map[string]string{
+		manifestEntryMetadataContentTypeKey: contentType,
+		manifestEntryMetadataFilenameKey:    fileName,
+	}
+
+	logger.Debugf("Uploading file Encrypt: %v Filename: %s Filehash: %s FileMtdt: %v",
+		encrypt, fileName, fr.String(), fileMtdt)
+
+	m, err := manifest.NewDefaultManifest(l, encrypt)
+	if err != nil {
+		logger.Debugf("bzz upload file: create manifest, file %q: %v", fileName, err)
+		logger.Errorf("bzz upload file: create manifest, file %q", fileName)
+		jsonhttp.InternalServerError(w, "could not create file manifest")
+		return
+	}
+
+	rootMetadata := map[string]string{
+		manifestWebsiteIndexDocumentSuffixKey: fileName,
+	}
+
+	err = m.Add(ctx, manifest.RootPath, manifest.NewEntry(swarm.ZeroAddress, rootMetadata))
```

Comment:
I wonder if some of the logic in this method can be simplified to use the `storeFile` (and `storeDir`?) from the directory upload method

Reply:
I thought about it, but with the single file upload, we need to add additional root metadata entry. In case of directories, we only add the file entry. Only part common is the code to create fileMetadata and file entry. I feel its not worth saving the few lines vs having to browse through a separate function for this.